test/compiling_tests.hpp

```diff
@@ -243,6 +243,7 @@ struct uncomparable_string {
 class null_constructible {
 public:
   null_constructible(int* p_)  : p(p_) {}
+  bool operator==(const std::nullptr_t&) const noexcept = delete;
```

Comment:
This is problematic. The idea is to ensure that an object is not implicitly constructed from a `nullptr` value. It is meant to emulate types like `std::string_view`. Explicitly deleting `operator==(std::nullptr_t)` makes this type an odd one, and not representative of the types that we want to avoid problems with.

Reply:
OK, I will try reverting this line, but it was necessary to resolve this issue with Visual Studio 2015. I suspect I am missing something in how `is_null_comparable` does its work.
An earlier attempt of mine made the `null_constructible` constructor `explicit`, before I
realized it was the wrong direction to take.

I also suspect it may be better in the short term for the test case not to be compiled,
so the GCC 4.x enabled build and `TROMPELOEIL_CONCAT` changes can merge sooner.
Consider the final state of this pull request a "part two of n" in that case, your original work
being "part one".

On the topic of the "problem types", `std::string_view` being one of them, I wondered while
working on this whether it would be worth creating a `compiling_tests_17.cpp` for test cases
specific to C++17.  A specific test with `std::string_view` might be one way to confirm
things haven't been broken (equivalently successfully fixed), rather that relying on
analogues of the types that are causing us problems.

I also wondered if it is worth encoding in `static_assert` statements the expected results
for Trompeloeil's type traits.  Early warning of compiler-specific differences might be useful.
We would also avoid testing them indirectly as we do now by relying on `print()` to test
`is_null_comparable`.

I further started to consider C++2a concepts as a means to model the expressions to have
hold for particular types, but it is too early in my learning about this language feature
to tell where it might lead.  Concept based checks might reasonably be placed in
`compiling_tests_20.cpp`.

Thoughts?

Reply:
Yes, yes, yes, yes.

Adding test suites for C++17 and C++20 is something we definitely should do. Not just ensuring that the C++11/14 constructs still builds. As for `std::string_view`, it was just the type that the problem was reported for. I'm sure there are lots of similar types written in C++11 code bases.

We should definitely do some static tests, maybe as static_asserts, for a lot of the metaprogramming hacks.

I'm less sure about adding C++17/20 constructs in the library. The reason is just that I hate the added complexity of having several interleaved code bases, separated by `#ifdef`. I much prefer to find solutions that works for all compilers and language versions. However, I can see two reasons for going there. The obvious one is to add new features to the library, that requires new functionality. That's a fairly harmless type of `#ifdef`. The other reason is if compilation time can be substantially reduced. For example, there's a fair amount of tag-dispatching going on, that could is C++17 be replaced with `if constexpr`. Likewise, in C++20, concepts could definitely be used. One such concept could be what this issue is about - null-comparable types. I feel that the perceived improvement for users must be quite noticeable for those changes, though, since it really adds a lot of complexity to already very complex code.